contracts/tokenbridge/ethereum/L1AtomicTokenBridgeCreator.sol

```diff
@@ -0,0 +1,454 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.4;
+
+import {L1GatewayRouter} from "./gateway/L1GatewayRouter.sol";
+import {L1ERC20Gateway} from "./gateway/L1ERC20Gateway.sol";
+import {L1CustomGateway} from "./gateway/L1CustomGateway.sol";
+import {L1WethGateway} from "./gateway/L1WethGateway.sol";
+import {ProxyAdmin} from "@openzeppelin/contracts/proxy/transparent/ProxyAdmin.sol";
+import {TransparentUpgradeableProxy} from "@openzeppelin/contracts/proxy/transparent/TransparentUpgradeableProxy.sol";
+import {L2AtomicTokenBridgeFactory, OrbitSalts} from "../arbitrum/L2AtomicTokenBridgeFactory.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {IInbox, IBridge, IOwnable} from "@arbitrum/nitro-contracts/src/bridge/IInbox.sol";
+import {AddressAliasHelper} from "../libraries/AddressAliasHelper.sol";
+import {Create2} from "@openzeppelin/contracts/utils/Create2.sol";
+import {BeaconProxyFactory, ClonableBeaconProxy} from "../libraries/ClonableBeaconProxy.sol";
+
+/**
+ * @title Layer1 token bridge creator
+ * @notice This contract is used to deploy token bridge on custom L2 chains.
+ * @dev Throughout the contract terms L1 and L2 are used, but those can be considered as base (N) chain and child (N+1) chain
+ */
+contract L1AtomicTokenBridgeCreator is Ownable {
```

Comment:
consider use a proxy for this contract, otherwise a re-deployment of the factory will change the canonicalL2FactoryAddress

Reply:
hm right. Also not sure do we want to have templates (L2 logic contracts deployed on L1) immutable? Currently factory owner can change it through `setTemplates` and that also impacts canonical addresses for L2 router and gateways

Reply:
Made L1Creator upgradeable, and L2 contracts use dummy contract as address seed before upgrading to actual implementation